File: src/statesman/commands/PointsCommand.java
package statesman.commands;

import statesman.Content;
import statesman.Interpreter;

public class PointsCommand extends Command {

    public enum PointsActionType {
        ADD, SUBTRACT, SET, CLEAR, LIST
    }

    public static final String ID_POINTS = "points";
    public static final String ID_POINTS_ADD = "add";
    public static final String ID_POINTS_SUB = "sub";
    public static final String ID_POINTS_SET = "set";
    public static final String ID_POINTS_CLEAR = "clear";
    public static final String ID_POINTS_LIST = "list";

    private PointsActionType _action;
    private int _value;

    private PointsCommand() {
        _action = null;
        _value = 0;
    }

    public PointsCommand(PointsActionType action, int value) {
        if (value < 0) {
            throw new IllegalArgumentException(
                    "Target points value cannot be less than zero");
        }
        if (action == null) {
            throw new IllegalArgumentException(
                    "Invalid value was passed to the action parameter");
        }
        _action = action;
        _value = value;
    }

    public PointsCommand(PointsActionType action) {
        this(action, 0);
    }

    private static Command _defaultInstance;

    public static Command getDefault() {
        if (_defaultInstance == null) {
            _defaultInstance = new PointsCommand();
        }
        return _defaultInstance;
    }

    @Override
    public Command fromText(String commandId, String[] arguments) {
        if (arguments.length < 2) {
            return null;
        }

        String action = arguments[1].trim();
        PointsActionType actionType = null;
        int actionValue = 0;
        if (arguments.length == 2) {
            if (action.equalsIgnoreCase(ID_POINTS_LIST)) {
                actionType = PointsActionType.LIST;
            } else if (action.equalsIgnoreCase(ID_POINTS_CLEAR)) {
                actionType = PointsActionType.CLEAR;
            }
        } else if (arguments.length == 3) {
            if (action.equalsIgnoreCase(ID_POINTS_ADD)) {
                actionType = PointsActionType.ADD;
            } else if (action.equalsIgnoreCase(ID_POINTS_SUB)) {
                actionType = PointsActionType.SUBTRACT;
            } else if (action.equalsIgnoreCase(ID_POINTS_SET)) {
                actionType = PointsActionType.SET;
            }
            try {
                actionValue = Integer.parseInt(arguments[2]);
            } catch (NumberFormatException ex) {
                // Second argument was not a number.
                return null;
            }
        }

        if (actionType == null) {
            return null;
        }
        return new PointsCommand(actionType, actionValue);
    }

    @Override
    public void execute() {
        switch (_action) {
        case ADD:
            Interpreter.setPoints(Interpreter.getPoints() + _value);
            break;
        case SUBTRACT:
            Interpreter.setPoints(Interpreter.getPoints() - _value);
            break;
        case SET:
            Interpreter.setPoints(_value);
            break;
        case CLEAR:
            Interpreter.setPoints(0);
            break;
        case LIST:
            System.out.printf(Content.getScript().getMessage("p_1"),
                    Interpreter.getPoints(),
                    Content.getScript().getMaxPoints());
            break;
        default:
            break;
        }
    }

}

File: src/Commands/PointsCommand.cs
namespace Statesman.Commands
{
    public class PointsCommand : Command
    {
        public enum PointsActionType
        {
            Add,
            Subtract,
            Set,
            Clear,
            List
        }

        public const string CommandPoints = "points";

        public const string kPointsAdd = "add";
        public const string kPointsSubtract = "sub";
        public const string kPointsSet = "set";
        public const string kPointsClear = "clear";
        public const string kPointsList = "list";

        private readonly PointsActionType _actionType;
        private readonly int _value;

        public PointsCommand(PointsActionType action, int value)
        {
            if (value < 0)
            {
                throw new ArgumentException("Target points value cannot be less than zero");
            }
            _actionType = action;
            _value = value;
        }

        public PointsCommand(PointsActionType action)
                : this(action, 0)
        {
        }

        public new static Command FromText(string commandName, string[] arguments)
        {
            if (commandName != CommandPoints || arguments.Length < 2)
            {
                return null;
            }

            string actionString = arguments[1].Trim().ToLowerInvariant();
            PointsActionType? actionType = null;
            int actionValue = 0;
            if (arguments.Length == 2)
            {
                if (actionString.Equals(kPointsList))
                {
                    actionType = PointsActionType.List;
                }
                else if (actionString.Equals(kPointsClear))
                {
                    actionType = PointsActionType.Clear;
                }
            }
            else if (arguments.Length == 3)
            {
                if (actionString.Equals(kPointsAdd))
                {
                    actionType = PointsActionType.Add;
                }
                else if (actionString.Equals(kPointsSubtract))
                {
                    actionType = PointsActionType.Subtract;
                }
                else if (actionString.Equals(kPointsSet))
                {
                    actionType = PointsActionType.Set;
                }
                
                if (!int.TryParse(arguments[2], out actionValue))
                {
                    return null;
                }
            }

            if (!actionType.HasValue)
            {
                return null;
            }
            return new PointsCommand(actionType.Value, actionValue);
        }

        public override void Execute()
        {
            switch (_actionType)
            {
                case PointsActionType.Add:
                    Interpreter.Points += _value;
                    break;
                case PointsActionType.Subtract:
                    Interpreter.Points -= _value;
                    break;
                case PointsActionType.Set:
                    Interpreter.Points = _value;
                    break;
                case PointsActionType.Clear:
                    Interpreter.Points = 0;
                    break;
                case PointsActionType.List:
                    Console.Write(
                            Content.Script.FindMessage("p_1"),
                            Interpreter.Points,
                            Content.Script.MaxPoints);
                    break;
                default:
                    break;
            }
        }
    }
}
